src/lib/server/lib.c

```diff
@@ -0,0 +1,230 @@
+ *
+ * @copyright 2022 The FreeRADIUS server project
+ */
+#include <freeradius-devel/server/lib.h>
+#include <freeradius-devel/server/main_config.h>
+#include <freeradius-devel/util/atexit.h>
+
+/*
+ *  Terminator for array of fr_lib_autoinit_t
+ */
+fr_lib_autoinit_t const fr_lib_autoinit_terminator = { .name = NULL };
+
+/*
+ *  Global list of libraries
+ */
+static fr_lib_list_t *lib_list;
+
+/** Parse the global config section for a library and call its init function
+ *
+ * @param[in] lib	to configure and initialise
+ * @return
+ *	- 0 on success
+ *	- -1 on failure
+ */
+static int lib_init_call(fr_lib_t *lib)
+{
+	CONF_SECTION	*global_cs, *cs;
+
+	/*
+	 *	Find relevant config section
+	 */
+	global_cs = cf_section_find(cf_root(main_config->root_cs), "global", NULL);
+	if (!global_cs) {
+		WARN("Missing global config section");
+		goto init;
+	}
+
+	cs = cf_section_find(global_cs, lib->name, NULL);
+	if (!cs) {
+		WARN("Missing \"%s\" subsection in global config", lib->name);
+		goto init;
+	}
+
+	if ((cf_section_rules_push(cs, lib->autoinit->config)) < 0 ||
+	    (cf_section_parse(lib, lib->autoinit->inst, cs) < 0)) {
+		cf_log_err(cs, "Failed evaluating configuration for libldap");
+		return -1;
+	}
+
+	/*
+	 *  Call the init callback if defined
+	 */
+init:
+	if (lib->autoinit->init && (lib->autoinit->init()) < 0) return -1;
+
+	lib->initialised = true;
+
+	return 0;
+}
+
+/** Initialise a list of libraries
+ *
+ * @param to_init	Array of autoinit structures detailing libraries to initialise
+ * @return
+ *	- 0 on succcess
+ *	- -1 on failure
+ */
+static int lib_autoinit(fr_lib_autoinit_t * const *to_init)
+{
+	fr_lib_autoinit_t * const *p = to_init;
+
+	for (; (*p)->name; p++) {
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
+
+		/*
+		 *  If the library is already initialised, just increase the reference count
+		 */
+		if ((lib) && (lib->initialised)) {
+			lib->instance_count ++;
+			return 0;
+		}
+
+		if (!lib) {
+			MEM(lib = talloc_zero(lib_list, fr_lib_t));
+			lib->name = talloc_typed_strdup(lib, (*p)->name);
+			lib->autoinit = *p;
+			fr_rb_insert(&lib_list->libs, lib);
+		}
+		lib->instance_count++;
+
+		/*
+		 *  If the main config parsing is not complete we can't initialise the library yet
+		 */
+		if (!main_config->root_cs) continue;
+
+		if (lib_init_call(lib) < 0) return -1;
+	}
+
+	return 0;
+}
+
+/** Callback for creation of "lib" symbols
+ *
+ */
+int fr_dl_lib_autoinit(UNUSED dl_t const *module, void *symbol, UNUSED void *user_ctx)
+{
+	if (lib_autoinit((fr_lib_autoinit_t **)symbol) < 0) return -1;
+
+	return 0;
+}
+
+/** Run free callbacks for external libraries no-longer in use
+ *
+ * @param[in] to_free	Array of autoinit structures detailing libraries to free
+ */
+static void lib_autofree(fr_lib_autoinit_t * const *to_free)
+{
+	fr_lib_autoinit_t * const *p = to_free;
+
+	for (; (*p)->name; p++) {
```

Comment:
Maybe stick the init on the same line.  Just makes it slightly clearer.

Reply:
Or p != fr_lib_autoinit_terminator. Ensures people actually use the terminator :)

Reply:
Maybe stick the init on the same line. Just makes it slightly clearer.